File: tools/pairtable.py
#!/usr/bin/env python3
"""Losslessly convert little-endian u16 pairs to readable text."""
import argparse
import struct
from pathlib import Path


def export_pairs(data):
    if not data or len(data) % 4:
        raise ValueError("pair table must contain whole nonempty u16 pairs")
    values = struct.unpack(f"<{len(data) // 2}H", data)
    return "".join(
        f"0x{values[index]:04x} 0x{values[index + 1]:04x}\n"
        for index in range(0, len(values), 2)
    )


def import_pairs(text):
    values = []
    for line_number, line in enumerate(text.splitlines(), 1):
        tokens = line.split("#", 1)[0].split()
        if not tokens:
            continue
        if len(tokens) != 2:
            raise ValueError(
                f"line {line_number}: expected exactly two values")
        try:
            pair = [int(token, 0) for token in tokens]
        except ValueError as error:
            raise ValueError(
                f"line {line_number}: invalid pair value") from error
        if any(not 0 <= value <= 0xffff for value in pair):
            raise ValueError(
                f"line {line_number}: pair value is out of range")
        values.extend(pair)
    if not values:
        raise ValueError("pair table is empty")
    return struct.pack(f"<{len(values)}H", *values)


def self_test():
    raw = struct.pack("<6H", 0, 1, 0x1234, 0xabcd, 0xffff, 2)
    if import_pairs(export_pairs(raw)) != raw:
        raise AssertionError("pair-table round-trip failed")
    print("self-test=ok")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--self-test", action="store_true")
    commands = parser.add_subparsers(dest="command")
    export = commands.add_parser("export")
    export.add_argument("input", type=Path)
    export.add_argument("output", type=Path)
    build = commands.add_parser("build")
    build.add_argument("input", type=Path)
    build.add_argument("output", type=Path)
    args = parser.parse_args()
    if args.self_test:
        self_test()
        if args.command is None:
            return
    if args.command is None:
        parser.error("a pair-table command is required")
    args.output.parent.mkdir(parents=True, exist_ok=True)
    if args.command == "export":
        args.output.write_text(export_pairs(args.input.read_bytes()))
    else:
        args.output.write_bytes(import_pairs(args.input.read_text()))


if __name__ == "__main__":
    main()

File: tools/pairtable.ts
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";

export function export_pairs(data: Uint8Array): string {
  if (data.length === 0 || data.length % 4 !== 0)
    throw new Error("pair table must contain whole nonempty u16 pairs");
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  let result = "";
  for (let offset = 0; offset < data.length; offset += 4)
    result += `0x${view.getUint16(offset, true).toString(16).padStart(4, "0")} 0x${view.getUint16(offset + 2, true).toString(16).padStart(4, "0")}\n`;
  return result;
}

export function import_pairs(text: string): Buffer {
  const values: number[] = [];
  for (const [index, line] of text.split(/\r?\n/).entries()) {
    const tokens = line.split("#", 1)[0].trim().split(/\s+/).filter(Boolean);
    if (tokens.length === 0) continue;
    if (tokens.length !== 2) throw new Error(`line ${index + 1}: expected exactly two values`);
    for (const token of tokens) {
      if (!/^[-+]?(?:0[xX][0-9a-fA-F]+|0[bB][01]+|0[oO][0-7]+|[0-9]+)$/.test(token))
        throw new Error(`line ${index + 1}: invalid pair value`);
      const number = Number(token);
      if (!Number.isInteger(number)) throw new Error(`line ${index + 1}: invalid pair value`);
      if (number < 0 || number > 0xffff) throw new Error(`line ${index + 1}: pair value is out of range`);
      values.push(number);
    }
  }
  if (values.length === 0) throw new Error("pair table is empty");
  const result = Buffer.alloc(values.length * 2);
  values.forEach((number, index) => result.writeUInt16LE(number, index * 2));
  return result;
}

export function self_test(): void {
  const raw = Buffer.alloc(12);
  [0, 1, 0x1234, 0xabcd, 0xffff, 2].forEach((value, index) => raw.writeUInt16LE(value, index * 2));
  if (!import_pairs(export_pairs(raw)).equals(raw)) throw new Error("pair-table round-trip failed");
  console.log("self-test=ok");
}

function main(args: string[]): void {
  if (args.includes("-h") || args.includes("--help")) {
    console.log("usage: pairtable.ts [--self-test] {export,build} INPUT OUTPUT");
    return;
  }
  if (args.includes("--self-test")) {
    self_test();
    args = args.filter((arg) => arg !== "--self-test");
    if (args.length === 0) return;
  }
  const [command, input, output] = args;
  if (!command || !input || !output || !["export", "build"].includes(command))
    throw new Error("a pair-table command is required");
  mkdirSync(dirname(output), { recursive: true });
  if (command === "export") writeFileSync(output, export_pairs(readFileSync(input)));
  else writeFileSync(output, import_pairs(readFileSync(input, "utf8")));
}

if (import.meta.main) main(process.argv.slice(2));
